### ISSUE TITLE
Safari and STP fail to run ~3000 url tests

### ISSUE THREAD

**Issue by graouts** [web-platform-tests/wpt#26393]
Some of the url WPT tests fail to run as reported by WPT.fyi:

* [url/a-element-origin-xhtml.xhtml](https://wpt.fyi/results/url/a-element-origin-xhtml.xhtml?label=experimental&label=master&aligned)
* [url/a-element-origin.html](https://wpt.fyi/results/url/a-element-origin.html?label=experimental&label=master&aligned)
* [url/a-element-xhtml.xhtml](https://wpt.fyi/results/url/a-element-xhtml.xhtml?label=experimental&label=master&aligned)
* [url/a-element.html](https://wpt.fyi/results/url/a-element.html?label=experimental&label=master&aligned)
* [url/url-constructor.html](https://wpt.fyi/results/url/url-constructor.html?label=experimental&label=master&aligned)
* [url/url-origin.html](https://wpt.fyi/results/url/url-origin.html?label=experimental&label=master&aligned)

Running them locally or via WPT.live show that they run fine when loaded directly in the browser (albeit with some failures, but there’s no harness-level issue).

[Comparing Safari 14 stable with STP 115](https://wpt.fyi/results/url?label=master&product=safari%5Bexperimental%5D&product=safari%5Bstable%5D&aligned) shows the same failures so it does not appear to be an STP-specific or particularly recent issue. I'm not sure how to dig further back to find the reported behavior with older Safari stable releases.

When running one of those tests locally, I get this output:

```
The trick was to add --channel stable to run WPT locally. Now the output is as follows:

wpt > ./wpt run safari url/a-element.html --channel stable
Running 1 tests in web-platform-tests

ERROR Unsupported test type print-reftest for product safari
  ▶ ERROR [expected OK] /url/a-element.html
  └   → 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/Users/antoine/Code/wpt/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 352, in run_func
    self.result = True, self.func(self.protocol, self.url, self.timeout)
  File "/Users/antoine/Code/wpt/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 458, in do_testharness
    done, rv = handler(result)
  File "/Users/antoine/Code/wpt/tools/wptrunner/wptrunner/executors/base.py", line 783, in call
    url, command, payload = result
TypeError: 'NoneType' object is not iterable
```

**Comment by graouts** [comment]
This is tracked by rdar://70996055 internally at Apple.

**Comment by jgraham** [comment]
So `result` here comes from `result = protocol.base.execute_script(self.script_resume % format_map, asynchronous=True)` in `executorwebdriver.py`. `script_resume` is https://github.com/web-platform-tests/wpt/blob/32c3c8c5d65295b74eb8415d09ca382c4df1dfe8/tools/wptrunner/wptrunner/executors/testharness_webdriver_resume.js which ends up calling https://github.com/web-platform-tests/wpt/blob/32c3c8c5d65295b74eb8415d09ca382c4df1dfe8/tools/wptrunner/wptrunner/testharnessreport.js#L24-L70 But if the data there is `None` something surprising happened; either the callback was called with `None` which doesn't look possible, or the command was cancelled for some reason (e.g. the page tried to navigate, the process crashed, etc.)

**Comment by graouts** [comment]
This could be due to me manually quitting Safari because the test completes and the window remains open.

**Comment by Hexcles** [comment]
I was able to reproduce the error after all (after rebooting my Macbook, which could totally just be a coincidence).

I did bisecting and found the culprit
https://github.com/web-platform-tests/wpt/commit/4e15edcc3ca51c72c3846133c7b175ecd94b3a9b#diff-760122a1234231092e5a1a2ca8655bffdda80ff57f58327f69823cf31423c39aR4638
**specifically** this input: `"http://example.com/\uD800\uD801\uDFFE\uDFFF\uFDD0\uFDCF\uFDEF\uFDF0\uFFFE\uFFFF?\uD800\uD801\uDFFE\uDFFF\uFDD0\uFDCF\uFDEF\uFDF0\uFFFE\uFFFF"`

Though I'm still not sure what's happening under the hood.

**Comment by gsnedders** [comment]
Containing lone surrogates most notably, which seems plausibly problematic with JSON.

**Comment by burg** [comment]
I investigated the safaridriver side of things. It looks the the decoding error comes from trying to decode a UTF-8 buffer as JSON using Foundation's +[NSJSONSerialization JSONObjectWithData:options:error:]. I'm pretty sure that the JSON string is being created via JSC's JSON.stringify in WebKit. I haven't investigated whether this is part of the test JS or as part of safaridriver's helper JS.

It's relevant to mention that JSON.stringify behavior around this changed in the past few years and all engines are aligned: https://github.com/tc39/proposal-well-formed-stringify

I'm ignorant when it comes to lone surrogates and other unicode problems, but my research suggests that these escapes are not legal JSON input, or that the result is implementation defined. I found some results here which suggest NSJSONSerialization rejects many weird surrogate related inputs when the result is implementation-defined.
http://seriot.ch/json/parsing.html#35

I'd like some clarification as to whether this input is a valid URL before considering a fix in safaridriver. If it's invalid input, then it should be removed from the test cases. Otherwise,
